Refactoring to support Workspace API inheritance

ncdc opened the pull request:
- Add "WithContext" functions to generated listers
- Add ability for developer to pass in the cert/key/token to use for the loopback client
- Fix a bug in CRD OpenAPI spec generation
- Replace kube-aggregator with our own "mini aggregator" that doesn't need APIServices any more. Aggregate discovery from generic control plane apiserver, apiextensions apiserver, and CRDs.
- Undo the contributed resources version hack (this functionality is now in kcp)

ncdc committed:
Add context to generated lister funcs

Signed-off-by: Andy Goldstein <andy.goldstein@redhat.com>

ncdc committed:
Regenerate listers

Signed-off-by: Andy Goldstein <andy.goldstein@redhat.com>

ncdc committed:
Print old & new stacks for same path registrations

ncdc committed:
Add cluster name support to meta accessor

ncdc committed:
HACK: let user set cert, key, token for loopback

ncdc committed:
HACK: fix crd openapi bug

Fix a bug in updateSpecLocked() where it was only updating the first
logical cluster's specs and then accidentally returning early.

ncdc committed:
HACK: discovery.GroupManager: expose Groups()

Add a GroupLister interface/Groups() method to the GroupManager
interface, allowing the mini aggregation server to aggregate API
discovery for multiple generic API servers.

ncdc committed:
HACK: mini aggregator

Replace full kube-aggregator with a new mini-aggregator that
only includes the generic control plane apiserver and the
apiextensions apiserver.

Aggregate discovery from generic control
plane, apiextensions, and custom resource definitions.

Aggregate OpenAPI as best as possible (it's per logical cluster,
but doesn't yet support API inheritance).

ncdc committed:
UNDO: contributed resources version hack

This is now handled in kcp

ncdc committed:
HACK: CRD discovery fixes

Tease APIResourcesForGroupVersion function out of
versionDiscoveryHandler.ServeHTTP so it can be used
by kcp when aggregating built in /api/v1 discovery data
with CRDs in /api/v1.

Fix issue where the core API group was being included
in CRD group discovery - this led to a bug where kubectl
tried to download /api/v1 discovery twice, and that broke
the client-side RESTMapper (because it saw each resource
twice).

ncdc commented on pkg/genericcontrolplane/aggregator.go:
```diff
@@ -14,287 +14,184 @@ See the License for the specific language governing permissions and
-) (*aggregatorapiserver.Config, error) {
+var (
+	// Scheme defines methods for serializing and deserializing API objects.
+	Scheme = runtime.NewScheme()
```
I now realize this is for the entire genericcontrolplane package. Might make sense to move to an aggregator subpackage?

sttts replied:
yes please

sttts replied:
Would move the whole aggregator code here into its own package

ncdc replied:
Added a new commit for this - PTAL

ncdc commented on pkg/genericcontrolplane/server.go:
```diff
@@ -97,25 +93,43 @@ func Run(completeOptions completedServerRunOptions, stopCh <-chan struct{}) erro
+
 // CreateServerChain creates the apiservers connected via delegation.
-func CreateServerChain(completedOptions completedServerRunOptions, stopCh <-chan struct{}) (*genericapiserver.GenericAPIServer, error) {
+func CreateServerChain(completedOptions completedServerRunOptions, stopCh <-chan struct{}) (*ServerChain, error) {
```
This change may not be ideal, but I definitely make use of CustomResourceDefinitions and GenericControlPlane in the kcp server code.

ncdc commented on staging/src/k8s.io/apimachinery/pkg/api/meta/interfaces.go:
```diff
@@ -72,6 +72,9 @@ type MetadataAccessor interface {
 	Continue(obj runtime.Object) (string, error)
 	SetContinue(obj runtime.Object, c string) error
 
+	ClusterName(obj runtime.Object) (string, error)
```
This isn't necessary in this PR - I was going to use it for something, but didn't end up needing it. I can back it out, or we can keep it.

stevekuznetsov replied:
We added this to meta.Accessor, so makes sense to have here as well

ncdc commented on staging/src/k8s.io/apimachinery/pkg/api/meta/meta.go:
```diff
@@ -375,6 +375,23 @@ func (resourceAccessor) SetContinue(obj runtime.Object, version string) error {
 	return nil
 }
 
+func (resourceAccessor) ClusterName(obj runtime.Object) (string, error) {
```
This isn't necessary in this PR - I was going to use it for something, but didn't end up needing it. I can back it out, or we can keep it.

ncdc commented on staging/src/k8s.io/apiserver/pkg/server/mux/pathrecorder.go:
```diff
@@ -103,10 +103,11 @@ func (m *PathRecorderMux) ListedPaths(clusterName string) []string {
+	stack := string(debug.Stack())
 	if existingStack, ok := m.pathStacks[path]; ok {
-		utilruntime.HandleError(fmt.Errorf("registered %q from %v", path, existingStack))
+		utilruntime.HandleError(fmt.Errorf("duplicate path registration of %q: original registration from %v\n\nnew registration from %v", path, existingStack, stack))
```
I should probably submit this separately to k/k - it's quite difficult to determine duplicate registrations if the error message only tells you where the new (duplicated) registration comes from.

sttts replied:
+1 for k/k

ncdc replied:
https://github.com/kubernetes/kubernetes/pull/106892

sttts replied:
tagged

ncdc commented on staging/src/k8s.io/apiserver/pkg/server/options/serving_with_loopback.go:
```diff
@@ -79,3 +99,32 @@ func (s *SecureServingOptionsWithLoopback) ApplyTo(secureServingInfo **server.Se
 	return nil
 }
+
+func (s *SecureServingOptionsWithLoopback) NewLoopbackClientCert() ([]byte, []byte, error) {
```
I wonder if it would be better to have this just set s.LoopbackCertPEM and s.LoopbackKeyPEM instead of returning the []bytes to the caller (and then making them set those 2 fields separately)?

ncdc replied:
@sttts WDYT?

sttts replied:
prefer the current factoring. In the new way proposed here it would be more like an on-demand defaulter. That's odd.

Haven't checked whether `NewLoopbackClientCert` uses anything from  `s`. I guess it does.

ncdc replied:
Ok, I can keep it like this.

stevekuznetsov commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go:
```diff
@@ -174,28 +197,40 @@ func (c completedConfig) New(delegationTarget genericapiserver.DelegationTarget)
+		crdLister: s.Informers.Apiextensions().V1().CustomResourceDefinitions().Lister(),
 		delegate:  delegateHandler,
 	}
+	s.versionDiscoveryHandler = versionDiscoveryHandler
```
nit: why this pattern of declaring a pointer, then setting the member, instead of `c.foo = &thing{}`?

ncdc replied:
Mostly evolution based on existing code and minimizing changes. Happy to go back and clean stuff up like this. Thanks for pointing it out.

ncdc replied:
Updated in a `CLEANUP` commit

stevekuznetsov commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_discovery.go:
```diff
@@ -42,63 +46,130 @@ func (r *versionDiscoveryHandler) ServeHTTP(w http.ResponseWriter, req *http.Req
-	}
-	r.discovery[clusterName][gv] = discoveryHandler
-}
+	// HACK: if we are adding resources in legacy scheme group through CRDs (KCP scenario)
```
nit: just :bomb: the commented out hack?

ncdc replied:
Should be gone now

stevekuznetsov commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_discovery.go:
```diff
@@ -42,63 +46,130 @@ func (r *versionDiscoveryHandler) ServeHTTP(w http.ResponseWriter, req *http.Req
 
-	if _, clusterExists := r.discovery[clusterName]; !clusterExists {
+	if !foundGroup || !foundVersion {
+		r.delegate.ServeHTTP(w, req)
```
what's the effect of doing this? what is our delegate?

stevekuznetsov replied:
or, why don't we have enough to 404 here ourselves?

ncdc replied:
I think this was the existing practice - if you can't handle something, just keep sending it down the line to the delegate. I'm guessing this would allow for consistent 404 responses? (delegate is `genericapiserver.NewEmptyDelegate()` in this particular case).

stevekuznetsov commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_discovery.go:
```diff
@@ -109,55 +180,148 @@ func (r *groupDiscoveryHandler) ServeHTTP(w http.ResponseWriter, req *http.Reque
 
-	if _, clusterExists := r.discovery[clusterName]; !clusterExists {
-		r.discovery[clusterName] = map[string]*discovery.APIGroupHandler{}
+	crds, err := r.crdLister.ListWithContext(ctx, labels.Everything())
```
nit: there seems to be a lot shared in the flow between these handlers, should we bother to DRY it out?

ncdc replied:
Yeah, I had that same thought, but didn't want to bother while I was hacking away. This code came from a controller, and it was unified because it set up both group and version discovery at the same time. But I ended up splitting it out into live handlers for each, and they're building different data sets. Maybe worthy of a TODO for now?

stevekuznetsov commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -275,76 +272,20 @@ func (r *crdHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+		utilruntime.HandleError(err)
 		responsewriters.ErrorNegotiated(
-			apierrors.NewInternalError(fmt.Errorf("error resolving resource: %v", err)),
+			apierrors.NewInternalError(fmt.Errorf("error resolving resource")),
```
Why is this good for the user?

ncdc replied:
This wasn't intentional. I think I had deleted this error handling originally during some refactoring, then ended up putting it back in, but I must have copied it from down below (where it doesn't have the %v). Good catch.

ncdc replied:
Added it back in a `CLEANUP` commit

stevekuznetsov commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -275,76 +272,20 @@ func (r *crdHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 		return
 	}
 
-	if cluster.Wildcard {
```
How does discovery for `*` work now?

ncdc replied:
This code was moved to kcp - no net changes.

stevekuznetsov commented on pkg/genericcontrolplane/aggregator.go:
```diff
@@ -14,287 +14,184 @@ See the License for the specific language governing permissions and
-// Package app does all of the work necessary to create a Kubernetes
-// APIServer by binding together the API, master and APIServer infrastructure.
-// It can be configured and called directly or via the hyperkube framework.
+// Package genericcontrolplane does TODO...
```
lol

ncdc replied:
Updated

stevekuznetsov commented on pkg/genericcontrolplane/aggregator.go:
```diff
@@ -14,287 +14,184 @@ See the License for the specific language governing permissions and
+	// TODO fix the server code to avoid this
+	metav1.AddToGroupVersion(Scheme, schema.GroupVersion{Version: "v1"})
+
+	// TODO: keep the generic API server from wanting this
```
Do we need e.g. WatchEvent etc?

ncdc replied:
This is currently only used to encode a small subset of response types. Definitely need to align on if we need a scheme here or if we should be using some other one.

sttts replied:
better call it discoveryScheme

ncdc replied:
Updated

stevekuznetsov commented on pkg/genericcontrolplane/aggregator.go:
```diff
@@ -14,287 +14,184 @@ See the License for the specific language governing permissions and
 
-	return aggregatorServer, nil
-}
+	combined := make([]metav1.APIGroup, 0, len(gcpGroups)+len(apiextensionsGroups)+len(crdGroups))
```
This is a bit curious use of `make()` - why?

ncdc replied:
We know the length in advance, so I thought we'd set it. What were you thinking of for this instead?

stevekuznetsov replied:
You also know you're about to fill it, so why set it to 0 and add a capacity?

ncdc replied:
Is it better to do `make([]foo, len(...))` and then set each element by index?

ncdc replied:
@stevekuznetsov did some Googling and learned that `copy` is more efficient for combining slices. Updated - PTAL.

stevekuznetsov reviewed (commented):
Can't speak to the legitimacy of the changes to core packages / how well they would review upstream, but all of this seems reasonable to me.

sttts commented on staging/src/k8s.io/apiserver/pkg/server/options/serving_with_loopback.go:
```diff
@@ -29,10 +31,15 @@ import (
 	*SecureServingOptions
+	LoopbackCertPEM []byte
+	LoopbackKeyPEM  []byte
+	LoopbackToken   string
```
godoc please

sttts replied:
then this can go upstream

ncdc replied:
Updated & restructured a bit - PTAL

sttts commented on pkg/genericcontrolplane/apiextensions.go:
```diff
@@ -86,10 +86,12 @@ func createAPIExtensionsConfig(
 			SharedInformerFactory: externalInformers,
 		},
 		ExtraConfig: apiextensionsapiserver.ExtraConfig{
-			CRDRESTOptionsGetter: apiextensionsoptions.NewCRDRESTOptionsGetter(etcdOptions),
```
can we split the discovery change and the client new func change? The later should be easier to upstream. The former will lead to a performance discussion (which we might win).

ncdc replied:
Definitely can do. Is it ok if we defer that until later, when we decide to start creating upstream PRs?

sttts replied:
This commit still has far too much of different things going on. It will catch us later. I think it's worth to split the New* funcs appart into their own commit.

ncdc replied:
Will do, thanks!

sttts commented on pkg/genericcontrolplane/aggregator.go:
```diff
@@ -14,287 +14,184 @@ See the License for the specific language governing permissions and
-		apiServices = append(apiServices, apiService)
+	// In order to reuse the kube-openapi API as much as possible, we
+	// register the OpenAPI service in the singlePathHandler
+	err = openAPIVersionedService.RegisterOpenAPIVersionedService("/openapi/v2", handler)
```
does this mean we merge openapi specs here in the hotpath?

ncdc replied:
(Note, this isn't net net to our forks - I've just moved code around.)

Yes, merging is done on the fly.

sttts replied:
can you add a TODO for optimization. Am not sure we want that long-term. At least we might want some caching. Merging is heavy.

ncdc commented on staging/src/k8s.io/apiserver/pkg/server/options/serving_with_loopback.go:
```diff
@@ -79,3 +127,34 @@ func (s *SecureServingOptionsWithLoopback) ApplyTo(secureServingInfo **server.Se
+	return certPem, keyPem, nil
+}
+
+// NewLoopbackClientConfig creates a loopback client *rest.Config for the given token and certificate.
```
@sttts note, this is not 100% identical to `(*secureServingInfo).NewLoopbackClientConfig()`. That function requires an active `Listener`. Should I try to refactor so we don't have duplicate similar methods?

sttts replied:
port could be zero here, can't it? We need that case for integration tests. Do those still run? Do we run those in CI ?

ncdc replied:
Good point... If it's 0, we can't create the client config until after the server has started, which won't work for kcp. I can have it return an error if the port is 0, and make the godoc explicit that it doesn't support 0, at least for now?

sttts replied:
yes, error + godoc is fine.

ncdc replied:
New commit added - PTAL

sttts commented on staging/src/k8s.io/apiserver/pkg/server/options/serving_with_loopback.go:
```diff
@@ -31,9 +32,24 @@ import (
+	// Token is the bearer token the loopback client uses to communicate with the server. if
+	// unset, will be generated when ApplyTo is called.
+	Token string
 }
```
:+1:

sttts commented:
Looks good. Can you squash your cleanups into the original commits?

ncdc commented:
Yes, will squash them, thanks

stevekuznetsov commented:
All of my small comments were addressed but I'll wait for Stefan to merge this

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_discovery.go:
```diff
@@ -117,11 +131,15 @@ func (r *versionDiscoveryHandler) ServeHTTP(w http.ResponseWriter, req *http.Req
+			if requestedVersion == v.Name {
+				subresources = v.Subresources
+				break
+			}
```
nit: didn't we have some helpers for these kind of version getters?

sttts replied:
actually it was used. Wouldn't it be the same as ignoring the err?

ncdc replied:
Yes, although for some reason I feel like ignoring an error is "bad". I can revert to how it was & ignore if you'd prefer?

sttts replied:
returning an error in that helper is actually bad style. Should be a bool.

ncdc replied:
I actually have a a better way to get the subresources that doesn't involve iterating through crd.spec.versions a second time - will push an update in a few

ncdc replied:
New commit added - PTAL

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/discovery/root.go:
```diff
@@ -33,12 +34,18 @@ import (
 }
 
+type GroupLister interface {
+	Groups(ctx context.Context, req *http.Request) ([]metav1.APIGroup, error)
```
that's an odd signature. Can you comment what ctx and req are used for?

ncdc replied:
I agree. I can get rid of req and move it back to the ServeHTTP flow. I was originally trying to make sure that we filled in the ServerAddressByClientCIDRs field everywhere we were returning discovery, including in our aggregation, but looking at what upstream does today, even that isn't consistent (some calls have it, some don't).

What do you think makes the most sense?

sttts replied:
am fine with either way, including just go-doc'ing what you just described about `ServerAddressByClientCIDRs`.

ncdc replied:
New commit added - PTAL

ncdc committed:
Overridable client/informers for CRD apiserver

Add NewClientFunc and NewInformerFactoryFunc for the
apiextensions-apiserver to allow composers to customize how the client
and shared informer factories are created.

Signed-off-by: Andy Goldstein <andy.goldstein@redhat.com>

ncdc committed:
HACK: make CRDs more extensible

Switch from controller-based discovery to controller-free discovery for
CRDs.

Remove logical cluster code from the crdHandler as it can be composed
from above (in kcp).

Signed-off-by: Andy Goldstein <andy.goldstein@redhat.com>

ncdc commented:
Merging as per discussion on both Slack & Meet!
